### PR TITLE
Autogenerate API documentation

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR is primarily adds the automatic generation of the API documentation for the entire Pyomo codebase to the Sphinx build.

This also silences all Sphinx warnings that were being generated

## Changes proposed in this PR:
- Add the API documentation under `Reference` / `Library Reference`
- Move the old library reference to `Reference` / `Topical Reference`
- Switch to using BiBTeX `alpha` format for all references (that are not references to Pyomo publications)
- Consolidated references into the bibliography
- Remove the (unused) Sphinx intersphinx reference 
- Remove the use of `sphinx.ext.autosectionlabel`
- Restored the working with abstract models" docs
- Extensive updates to docstrings to resolve Sphinx warnings
- Pyomo code
  - cleaned `ConfigDict` numpydoc generation (improve references, resolve sphinx warnings)
  - Add an `IntEnum` wrapper to rewrite some docstrings to silence Sphinx warnings
  - remove use of "bare" absolute imports in Pyomo `__init__.py` files (resolves Sphinx autosummary issue)
  - Prevent importing Qt when building docs
  - Switch from docstrings to decorators for defining PLY grammars

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/kernel/__init__.py [hunk 1]
```diff
@@ -59,20 +59,22 @@
...
-import pyomo.core.kernel.matrix_constraint
-import pyomo.core.kernel.parameter
-import pyomo.core.kernel.expression
-import pyomo.core.kernel.objective
-import pyomo.core.kernel.sos
-import pyomo.core.kernel.suffix
-import pyomo.core.kernel.block
-import pyomo.core.kernel.piecewise_library
-import pyomo.core.kernel.set_types
+from . import (
```
Comment (mrmundt) [inline pyomo/core/kernel/__init__.py:62]: I again register my relative path complaint. And now that it's for the third time I will express why:

- IDE and Tooling Support: Many IDEs and static analysis tools have better support for resolving absolute imports.
- Flexibility and Scalability: Absolute imports remain stable and clear regardless of how deep the package structure goes.

File: pyomo/core/kernel/piecewise_library/__init__.py [hunk 2]
```diff
@@ -9,6 +9,4 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.core.kernel.piecewise_library.util
-import pyomo.core.kernel.piecewise_library.transforms
-import pyomo.core.kernel.piecewise_library.transforms_nd
+from . import util, transforms, transforms_nd
```
Comment (mrmundt) [inline pyomo/core/kernel/piecewise_library/__init__.py:12]: Relative path complaint.

File: pyomo/core/kernel/piecewise_library/transforms.py [hunk 3]
```diff
@@ -15,8 +15,9 @@
 mixed-integer problem formulation. Reference::
 
   Mixed-Integer Models for Non-separable Piecewise Linear
-Optimization: Unifying framework and Extensions (Vielma,
-Nemhauser 2008)
+  Optimization: Unifying framework and Extensions (Vielma,
+  Nemhauser 2008)
```
Comment (mrmundt) [inline pyomo/core/kernel/piecewise_library/transforms.py:19]: Should this be put in the bibliography?

File: pyomo/core/kernel/piecewise_library/transforms_nd.py [hunk 4]
```diff
@@ -15,8 +15,9 @@
 mixed-integer problem formulation. Reference::
 
   Mixed-Integer Models for Non-separable Piecewise Linear
-Optimization: Unifying framework and Extensions (Vielma,
-Nemhauser 2008)
+  Optimization: Unifying framework and Extensions (Vielma,
+  Nemhauser 2008)
```
Comment (mrmundt) [inline pyomo/core/kernel/piecewise_library/transforms_nd.py:19]: Bibliography?

File: pyomo/core/plugins/__init__.py [hunk 5]
```diff
@@ -11,4 +11,4 @@
 
 
 def load():
-    import pyomo.core.plugins.transform
+    from . import transform
```
Comment (mrmundt) [inline pyomo/core/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/core/plugins/transform/__init__.py [hunk 6]
```diff
@@ -9,18 +9,17 @@
...
-# import pyomo.core.plugins.transform.equality_transform
-import pyomo.core.plugins.transform.nonnegative_transform
-import pyomo.core.plugins.transform.radix_linearization
-import pyomo.core.plugins.transform.discrete_vars
-
-# import pyomo.core.plugins.transform.util
-import pyomo.core.plugins.transform.add_slack_vars
-import pyomo.core.plugins.transform.scaling
-import pyomo.core.plugins.transform.logical_to_linear
+from . import (
```
Comment (mrmundt) [inline pyomo/core/plugins/transform/__init__.py:12]: Relative path complaint.

File: pyomo/dae/plugins/__init__.py [hunk 7]
```diff
@@ -11,5 +11,4 @@
 
 
 def load():
-    import pyomo.dae.plugins.colloc
-    import pyomo.dae.plugins.finitedifference
+    from . import colloc, finitedifference
```
Comment (mrmundt) [inline pyomo/dae/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/dataportal/__init__.py [hunk 8]
```diff
@@ -9,7 +9,7 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.dataportal.parse_datacmds
+from . import parse_datacmds
```
Comment (mrmundt) [inline pyomo/dataportal/__init__.py:12]: Relative path complaint.

File: pyomo/duality/__init__.py [hunk 9]
```diff
@@ -9,4 +9,4 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.duality.collect
+from . import collect
```
Comment (mrmundt) [inline pyomo/duality/__init__.py:12]: Relative path complaint.

File: pyomo/gdp/plugins/__init__.py [hunk 10]
```diff
@@ -11,15 +11,17 @@
...
-    import pyomo.gdp.plugins.gdp_var_mover
-    import pyomo.gdp.plugins.cuttingplane
-    import pyomo.gdp.plugins.fix_disjuncts
-    import pyomo.gdp.plugins.partition_disjuncts
-    import pyomo.gdp.plugins.between_steps
-    import pyomo.gdp.plugins.multiple_bigm
-    import pyomo.gdp.plugins.transform_current_disjunctive_state
-    import pyomo.gdp.plugins.bound_pretransformation
-    import pyomo.gdp.plugins.binary_multiplication
+    from . import (
```
Comment (mrmundt) [inline pyomo/gdp/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/mpec/plugins/__init__.py [hunk 11]
```diff
@@ -11,10 +11,4 @@
...
 
 def load():
-    import pyomo.mpec.plugins.mpec1
-    import pyomo.mpec.plugins.mpec2
-    import pyomo.mpec.plugins.mpec3
-    import pyomo.mpec.plugins.mpec4
-    import pyomo.mpec.plugins.solver1
-    import pyomo.mpec.plugins.solver2
-    import pyomo.mpec.plugins.pathampl
+    from . import mpec1, mpec2, mpec3, mpec4, solver1, solver2, pathampl
```
Comment (mrmundt) [inline pyomo/mpec/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/neos/plugins/__init__.py [hunk 12]
```diff
@@ -11,5 +11,4 @@
 
 
 def load():
-    import pyomo.neos.plugins.NEOS
-    import pyomo.neos.plugins.kestrel_plugin
+    from . import NEOS, kestrel_plugin
```
Comment (mrmundt) [inline pyomo/neos/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/network/plugins/__init__.py [hunk 13]
```diff
@@ -11,4 +11,4 @@
 
 
 def load():
-    import pyomo.network.plugins.expand_arcs
+    from . import expand_arcs
```
Comment (mrmundt) [inline pyomo/network/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/opt/parallel/__init__.py [hunk 14]
```diff
@@ -15,5 +15,4 @@
     SolverManagerFactory,
     AsynchronousSolverManager,
 )
-import pyomo.opt.parallel.manager
-import pyomo.opt.parallel.local
+from . import manager, local
```
Comment (mrmundt) [inline pyomo/opt/parallel/__init__.py:18]: Relative path complaint.

File: pyomo/opt/plugins/__init__.py [hunk 15]
```diff
@@ -11,6 +11,4 @@
 
 
 def load():
-    import pyomo.opt.plugins.driver
-    import pyomo.opt.plugins.res
-    import pyomo.opt.plugins.sol
+    from . import driver, res, sol
```
Comment (mrmundt) [inline pyomo/opt/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/repn/beta/__init__.py [hunk 16]
```diff
@@ -9,4 +9,4 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.repn.beta.matrix
+from . import matrix
```
Comment (mrmundt) [inline pyomo/repn/beta/__init__.py:12]: Relative path complaint.

File: pyomo/repn/plugins/__init__.py [hunk 17]
```diff
@@ -11,15 +11,16 @@
...
-    import pyomo.repn.plugins.cpxlp
-    import pyomo.repn.plugins.ampl
-    import pyomo.repn.plugins.baron_writer
-    import pyomo.repn.plugins.mps
-    import pyomo.repn.plugins.gams_writer
-    import pyomo.repn.plugins.lp_writer
-    import pyomo.repn.plugins.nl_writer
-    import pyomo.repn.plugins.standard_form
-
+    from . import (
```
Comment (mrmundt) [inline pyomo/repn/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/scripting/__init__.py [hunk 18]
```diff
@@ -9,5 +9,4 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.scripting.pyomo_command
-import pyomo.scripting.util
+from . import pyomo_command, util
```
Comment (mrmundt) [inline pyomo/scripting/__init__.py:12]: Relative path complaint.

File: pyomo/scripting/plugins/__init__.py [hunk 19]
```diff
@@ -11,8 +11,4 @@
 
 
 def load():
-    import pyomo.scripting.plugins.convert
-    import pyomo.scripting.plugins.solve
-    import pyomo.scripting.plugins.download
-    import pyomo.scripting.plugins.build_ext
-    import pyomo.scripting.plugins.extras
+    from . import convert, solve, download, build_ext, extras
```
Comment (mrmundt) [inline pyomo/scripting/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/solvers/plugins/__init__.py [hunk 20]
```diff
@@ -11,5 +11,4 @@
 
 
 def load():
-    import pyomo.solvers.plugins.converter
-    import pyomo.solvers.plugins.solvers
+    from . import converter, solvers
```
Comment (mrmundt) [inline pyomo/solvers/plugins/__init__.py:14]: Relative path complaint.

File: pyomo/solvers/plugins/converter/__init__.py [hunk 21]
```diff
@@ -9,6 +9,4 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-import pyomo.solvers.plugins.converter.ampl
-import pyomo.solvers.plugins.converter.glpsol
-import pyomo.solvers.plugins.converter.model
+from . import ampl, glpsol, model
```
Comment (mrmundt) [inline pyomo/solvers/plugins/converter/__init__.py:12]: Relative path complaint.

File: pyomo/solvers/plugins/solvers/__init__.py [hunk 22]
```diff
@@ -10,24 +10,26 @@
...
-import pyomo.solvers.plugins.solvers.gurobi_persistent
-import pyomo.solvers.plugins.solvers.cplex_direct
-import pyomo.solvers.plugins.solvers.cplex_persistent
-import pyomo.solvers.plugins.solvers.GAMS
-import pyomo.solvers.plugins.solvers.mosek_direct
-import pyomo.solvers.plugins.solvers.mosek_persistent
-import pyomo.solvers.plugins.solvers.xpress_direct
-import pyomo.solvers.plugins.solvers.xpress_persistent
-import pyomo.solvers.plugins.solvers.SAS
+from . import (
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/__init__.py:13]: Relative path complaint.

File: pyomo/solvers/tests/models/__init__.py [hunk 23]
```diff
@@ -9,40 +9,34 @@
...
-import pyomo.solvers.tests.models.MIQP_simple
-
-import pyomo.solvers.tests.models.QCP_simple
-
-import pyomo.solvers.tests.models.QP_constant_objective
-import pyomo.solvers.tests.models.QP_simple
-
-import pyomo.solvers.tests.models.SOS1_simple
-import pyomo.solvers.tests.models.SOS2_simple
+from . import (
```
Comment (mrmundt) [inline pyomo/solvers/tests/models/__init__.py:12]: Relative path complaint.

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/core/base/logical_constraint.py [hunk 24]
```diff
@@ -47,17 +47,24 @@ class LogicalConstraintData(ActiveComponentData):
...
+            The Pyomo expression for this logical constraint
 
     Private class attributes:
-        _component      The logical constraint component.
-        _active         A boolean that indicates whether this data is active
+        _component
+            The logical constraint component.
+        _active
+            A boolean that indicates whether this data is active
+
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/base/logical_constraint.py:67]: I changed it elsewhere as part of squashing Sphinx warnings.  I am trying NOT to update too much documentation in the PR (this is already complex enough).  In addition, the `logical_*` classes are "next" up to be rewritten to bring them in line with the numeric classes they were originally copied from (so the docs will be updated when we update the class).

File: pyomo/util/subsystems.py [hunk 25]
```diff
@@ -286,26 +286,33 @@ class ParamSweeper(TemporarySubsystemManager):
...
+       :skipif: not glpk_available
+
+       model = pyo.ConcreteModel()
+       model.v1 = pyo.Var()
+       model.v2 = pyo.Var()
+       model.c = pyo.Constraint(expr=model.v2 - model.v1 >= 0.1)
+       model.o = pyo.Objective(expr=model.v1 + model.v2)
+       solver = pyo.SolverFactory('glpk')
+       input_vars = [model.v1]
+       n_scen = 2
```
Comment (mrmundt) [inline pyomo/util/subsystems.py:299]: Not relevant for this PR but marking for posterity - what is `n_scen`?

**Review by mrmundt — APPROVED** [review]
Yes yes, I am pleased.

File: pyomo/util/subsystems.py (hunk 25 above)
Comment (jsiirola) [reply pyomo/util/subsystems.py:299]: I believe "number of scenarios"

**Pull request merged**